### ISSUE TITLE
All cities walkability should be based on weighted average of local 'all cities' walkability

### ISSUE THREAD

**Issue by carlhiggs** [healthysustainablecities/global-indicators#26]
As per discussion with Shirley on Basecamp [here](https://3.basecamp.com/3662734/buckets/11779922/messages/2465025799), I believe we have identified that the city level walkability indices are being calculated incorrectly.

I think it would be useful to keep the initial important theoretical discussion on Basecamp so Billie, Jim and others are on board.  Then, we can have code related discussion here about implementation.

**Example of the issue**
For example, for Bangkok the 'all_cities_walkability' score of -1.97 is the sum of the marginal population weighted scores for intersection density, population density and daily living.  This is incorrect and not a valid measure of walkability, as it doesn't account for the local neighbourhood interaction of these variables.  

Instead, 'all_cities_walkability' should be based on the population weighted mean of the hex-level estimates of 'all cities walkability'.  Based on my quick checks, the actual between city Walkability for Bangkok's population should be 0.72, which makes more sense.

I'll have limited scope to look at this in the next few days; in the first instance, it will be interesting to hear if others agree with this interpretation.

**Comment by shiqin-liu** [comment]
Thanks, Carl. So based on what you have said in basecamp (if I understand it correctly), we need to aggregate hex-level walkability relative to all cities based, weighted by population est at the hex level. 
As you noted: 

> What we need to change
> The component variables are already there (population estimate and all_cities_walkability estimate at hex level); all we need to amend is the final aggregation to ensure the final measure is based on a population weighted average of the hex level estimates.
> 

I think, I think an exisitng function [calc_city(gpkg_hex_250m, city, gpkg_input, config, gpkg_output)](https://github.com/gboeing/global-indicators/blob/6f172d39fd9ccf2d62e3ec6de4dce48ac4accbf3/process/sv_aggr.py#L192) may do the job:

To illustrate, if we want to calculate the correct city-level walkability and daily living, we can revise the function to:
```
def calc_city(gpkg_hex_250m, city, gpkg_input, config, gpkg_output):
    """
    Calculate population-weighted city-level indicators across all cities,
    and save to output geopackage

    These indicators include:
        "all_cities_pop_z_daily_living",
        "all_cities_walkability"

    Parameters
    ----------
    gpkg_hex_250m: str
        file path of across-city hexagon-level indicators
    city: str
        name of a city
    gpkg_input: str
        file path of input geopackage
    config: dict
        dict read from configeration file
    gpkg_output: str
        file path of output geopackage

    Returns
    -------
    list, list of GeoDataFrame
    """
    gdf_hex = gpd.read_file(gpkg_hex_250m, layer=city)

    gdf_hex_origin = gpd.read_file(gpkg_input, layer=config["hex250"])
    gdf_study_region = gpd.read_file(gpkg_input,
                                     layer=config["urban_study_region"])
    # join pop_est from original hex to processed hex
    gdf_hex = gdf_hex.join(gdf_hex_origin.set_index('index'),
                           on='index',
                           how='left',
                           rsuffix='_origin')

    gdf_study_region[
        config['city_fieldNames']['urban_sample_point_count']] = gdf_hex[
            config['hex_fieldNames']['urban_sample_point_count']].sum()

    # hex-level field names from input gpkg
    fieldNames = [
        config['hex_fieldNames']['all_cities_z_daily_living'],
        config['hex_fieldNames']['all_cities_walkability']
    ]

    # new file names for population-weighted city-level indicators
    fieldNames_new = [
        config['city_fieldNames']['all_cities_pop_z_daily_living'],
        config['city_fieldNames']['all_cities_walkability']
    ]

    gdf_study_region = aggregation_city(gdf_hex, gdf_study_region, config,
                                        list(zip(fieldNames, fieldNames_new)))

    gdf_study_region.to_file(gpkg_output, layer=city, driver='GPKG')
    return gdf_study_region
```

Because from my understanding, `aggregation_city(input_gdf, out_gdf, config, fieldNames)` do the job to generate population weighted city-level indicators:
```
def aggregation_city(input_gdf, out_gdf, config, fieldNames):
    """
    Aggregating hexagon level indicators to city level by weighted population


    Parameters
    ----------
    input_gdf: GeoDataFrame
        GeoDataFrame of input hexagon
    out_gdf: GeoDataFrame
        GeoDataFrame of output city
    config: dict
        dict read from configeration file
    fieldNames: list(zip)
        fieldNames of hex-level and city-level indicators

    Returns
    -------
    GeoDataFrame
    """
    # loop over each indicators field names of input and output gdf
    for field in fieldNames:
        # calculate the population weighted indicators based on input hexagon layer
        # sum to aggregate up to the city level
        out_gdf[field[1]] = (input_gdf[config['pop_est']] * input_gdf[field[0]]
                             ).sum() / (input_gdf[config['pop_est']].sum())
    return out_gdf
```

Do you think the above achieve what we have discussed on basecamp?

**Comment by carlhiggs** [comment]
Hi Shirley,

Thanks - I think in theory it should! 

In practice, I find the verbosity of the nested config file parameter names repeated throught SV functions in conjunction with their very non-linear and non-transparent function names make it hard for me to read (e.g. i had to read all the way to the very last line to confirm that it appears that they use the correct weighting variable ... but having said that, I'm still unsure as looking in one of the config files, i can't see a 'pop_est' parameter to confirm this is using the correct parameter!  So I'm missing something -- and I think the way the code is written doesn't help.  Although, its always hard with "someone else's code"; so the problem is in part me).

As per the Basecamp post, I also think we need to population weight the other variables (and I am assuming - but haven't verified - that the population percentages do this already.  I'm trying to catch up with other projects over the next few days (while dealing with COVID19 workplace contingency plans) so likely won't look at this (or the other issues we have noted) until mid way through next week sorry.  I hope all is well with you all re current global happenings.

**Comment by shiqin-liu** [comment]
Hi, Carl,

I share the same concern with you while working on the docstrings and commenting. Current documentation makes the logic very hard to follow. The "pop_est" here is the hex-level population estimation within each specific city (in city-specific input geopackadge pop_ghs layer), so I think it does the job for calculating the "pop weighted" measures. 

I will revise our codes following this logic then, and also retaining the other variables (your explanation basecamp has been very helpful). I will aim to share more on this before our spatial team meeting next week. Thanks for still responding in time within your busy schedule!

**Comment by shiqin-liu** [comment]
A quick check with the above code gives me the expected result of walkability between cities for Bangkok (0.72):
<img width="579" alt="Screen Shot 2020-03-12 at 3 23 41 PM" src="https://user-images.githubusercontent.com/46453772/76559523-b773eb00-6475-11ea-97e3-06406d9e76c7.png">

**Comment by carlhiggs** [comment]
This issue has been addressed by the full distance function branches which have since been merged into master; so, marking as closed.  Thanks!
